src/clev2er/algorithms/seaice/alg_pulse_peakiness.py

```diff
@@ -0,0 +1,176 @@
+""" clev2er.algorithms.seaice.pulse_peakiness.py
+
+    Algorithm class module, used to implement a single chain algorithm
+
+    #Description of this Algorithm's purpose
+    
+    Calculates the pulse peakiness of each waveform, which is used later to discriminate
+    between specular and diffuse waveforms. Described in https://doi.org/10.1016/j.asr.2017.10.051
+    
+    #Main initialization (init() function) steps/resources required
+
+    None
+
+    #Main process() function steps
+    
+    Define peakiness calculation
+    Apply to every waveform
+    Add to shared_dict
+    
+    #Contribution to shared_dict
+
+    shared_dict["pulse_peakiness"] (np.array[float]) : the pulse peakiness of each waveform
+
+    #Requires from shared_dict
+
+    shared_dict["waveform"]
+    
+"""
+
+from typing import Tuple
+
+import numpy as np
+from codetiming import Timer  # used to time the Algorithm.process() function
+from netCDF4 import Dataset  # pylint:disable=no-name-in-module
+
+from clev2er.algorithms.base.base_alg import BaseAlgorithm
+
+# each algorithm shares some common class code, so pylint: disable=duplicate-code
+
+
+class Algorithm(BaseAlgorithm):
+    """CLEV2ER Algorithm class
+
+    contains:
+         .log (Logger) : log instance that must be used for all logging, set by BaseAlgorithm
+         .config (dict) : configuration dictionary, set by BaseAlgorithm
+         - functions that need completing:
+         .init() : Algorithm initialization function (run once at start of chain)
+         .process(l1b,shared_dict) : Algorithm processing function (run on every L1b file)
+         .finalize() : Algorithm finalization/closure function (run after all chain
+                       processing completed)
+
+    Inherits from BaseAlgorithm which handles interaction with the chain controller run_chain.py
+
+    """
+
+    def init(self) -> Tuple[bool, str]:
+        """Algorithm initialization function
+
+        Add steps in this function that are run once at the beginning of the chain
+        (for example loading a DEM or Mask)
+
+        Returns:
+            (bool,str) : success or failure, error string
+
+        Test for KeyError or OSError exceptions and raise them if found
+        rather than just returning (False,"error description")
+
+        Raises:
+            KeyError : for keys not found in self.config
+            OSError : for any file related errors
+
+        Note:
+        - retrieve required config data from self.config dict
+        - log using self.log.info(), or self.log.error() or self.log.debug()
+
+        """
+        self.alg_name = __name__
+        self.log.info("Algorithm %s initializing", self.alg_name)
+
+        # --- Add your initialization steps below here ---
+
+        self.noise_floor_start = self.config["alg_pulse_peakiness"]["noise_floor_start_bin"]
+        self.noise_floor_end = self.config["alg_pulse_peakiness"]["noise_floor_end_bin"]
+
+        def peakiness(waveform):
```

Comment:
A few things with this function.

It should have a docstring that defines its parameter arguments and returns (using the Google style docstring format)

As this is a general function to calculate peakiness, I think this would better placed as a utility function outside the algorithm and imported. ie in $CLEV2ER_BASE_DIR/src/clev2er/utils/waveforms/peakiness.py so that it can be reused by other algorithms in this or other projects.

Reply:
Also please add type hints to the function

Reply:
The function is now clev2er/utils/waveforms/peakiness.py, and now has a docstring for the function, the module and type hints for args and returned values.

I've also done this before, will go back and correct the other times I've included a function like that.